osu.Game.Rulesets.Taiko/UI/DrawableTaikoMascot.cs

```diff
@@ -0,0 +1,105 @@
+using osu.Game.Graphics.Containers;
+
+namespace osu.Game.Rulesets.Taiko.UI
+{
+    public sealed class DrawableTaikoMascot : BeatSyncedContainer
+    {
+        private static TaikoMascotTextureAnimation idleDrawable, clearDrawable, kiaiDrawable, failDrawable;
+        private EffectControlPoint lastEffectControlPoint;
+        private TaikoMascotAnimationState state;
+
+        public Bindable<TaikoMascotAnimationState> PlayfieldState;
+
+        /// <summary>
+        /// Determines if there should be no "state logic", intended for testing.
+        /// </summary>
+        public bool Dumb { get; set; }
+
+        public TaikoMascotAnimationState State
+        {
+            get => state;
+            set
+            {
+                state = value;
+
+                foreach (var child in InternalChildren)
+                    child.Hide();
+
+                var drawable = getStateDrawable(State);
+
+                drawable?.Show();
+            }
+        }
+
+        public DrawableTaikoMascot(TaikoMascotAnimationState startingState = TaikoMascotAnimationState.Idle)
+        {
+            RelativeSizeAxes = Axes.Both;
+            PlayfieldState = new Bindable<TaikoMascotAnimationState>();
+            PlayfieldState.BindValueChanged(b =>
+            {
+                if (lastEffectControlPoint != null)
+                    State = getFinalAnimationState(lastEffectControlPoint, b.NewValue);
+            });
+
+            State = startingState;
+        }
+
+        private TaikoMascotTextureAnimation getStateDrawable(TaikoMascotAnimationState state)
+        {
+            return state switch
+            {
+                TaikoMascotAnimationState.Idle => idleDrawable,
+                TaikoMascotAnimationState.Clear => clearDrawable,
+                TaikoMascotAnimationState.Kiai => kiaiDrawable,
+                TaikoMascotAnimationState.Fail => failDrawable,
+                _ => null
+            };
+        }
+
+        private TaikoMascotAnimationState getFinalAnimationState(EffectControlPoint effectPoint, TaikoMascotAnimationState playfieldState)
+        {
+            if (playfieldState == TaikoMascotAnimationState.Fail)
+                return playfieldState;
+
+            return effectPoint.KiaiMode ? TaikoMascotAnimationState.Kiai : TaikoMascotAnimationState.Idle;
+        }
+
+        [BackgroundDependencyLoader]
+        private void load(TextureStore textures)
+        {
+            InternalChildren = new[]
+            {
+                idleDrawable = new TaikoMascotTextureAnimation(TaikoMascotAnimationState.Idle),
+                clearDrawable = new TaikoMascotTextureAnimation(TaikoMascotAnimationState.Clear),
+                kiaiDrawable = new TaikoMascotTextureAnimation(TaikoMascotAnimationState.Kiai),
+                failDrawable = new TaikoMascotTextureAnimation(TaikoMascotAnimationState.Fail),
+            };
+
+            // making sure we have the correct sprite set
+            State = state;
```

Comment:
This is very weirdly written. I'd say invoking a property's setter with its own backing field as the value, solely to trigger the setter's side effects, is bad practice. The way forward here for me would be:

* to extract the sprite visibility toggling code to a method,
* to remove it from the setter entirely,
* and finally to move the `BindValueChanged` subscription from the ctor to here, and call it with the extracted method and `runOnceImmediately: true` to get the effect you intended to get by doing this hack.

I *think* it's equivalent but I haven't actually gone to check so buyer beware.

Reply:
> and finally to move the BindValueChanged subscription from the ctor to here, and call it with the extracted method and runOnceImmediately: true to get the effect you intended to get by doing this hack.

The last part seems to me you're mixing up intentions of my code (which I can't blame you for obviously).
The PlayfieldState bindable and State property have different functions.
PlayfieldState is used by the taiko playfield to set current information on whether the input was judged good or bad.
State itself was for setting the appearance of DrawableTaikoMascot, but even that was meant to be set by the `DrawableTaikoMascot` itself (see `OnNewBeat`) (reason why it is public is that it's accessible for testing).

Reply:
While I did miss that distinction the same conclusion sort of applies - ideally you'd need just one of these and the way forward I proposed earlier sounds like a possible way out.